src/Extension/Extension.php

```diff
@@ -0,0 +1,252 @@
+     */
+    protected $id;
+    /**
+     * The directory of this extension.
+     *
+     * @var string
+     */
+    protected $path;
+
+    /**
+     * Composer json of the package.
+     *
+     * @var array
+     */
+    protected $composerJson;
+
+    /**
+     * Whether the extension is installed.
+     *
+     * @var bool
+     */
+    protected $installed = true;
+
+    /**
+     * The installed version of the extension.
+     *
+     * @var string
+     */
+    protected $version;
+
+    /**
+     * Whether the extension is enabled.
+     *
+     * @var bool
+     */
+    protected $enabled = false;
+
+    /**
+     * @param       $path
+     * @param array $composerJson
+     */
+    public function __construct($path, $composerJson)
+    {
+        $this->id           = end(explode('/', $path));
+        $this->path         = $path;
+        $this->composerJson = $composerJson;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __get($name)
+    {
+        return $this->composerJsonAttribute(Str::snake($name, '-'));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __isset($name)
+    {
+        return isset($this->{$name}) || $this->composerJsonAttribute(Str::snake($name, '-'));
+    }
+
+
+    /**
+     * Dot notation getter for composer.json attributes.
+     *
+     * @see https://laravel.com/docs/5.1/helpers#arrays
+     *
+     * @param $name
+     * @return mixed
+     */
+    public function composerJsonAttribute($name)
+    {
+        return Arr::get($this->composerJson, $name);
+    }
+
+    /**
+     * @param boolean $installed
+     * @return Extension
+     */
+    public function setInstalled($installed)
+    {
+        $this->installed = $installed;
+
+        return $this;
+    }
+
+    /**
+     * @return boolean
+     */
+    public function isInstalled()
+    {
+        return $this->installed;
+    }
+
+    /**
+     * @param string $version
+     * @return Extension
+     */
+    public function setVersion($version)
+    {
+        $this->version = $version;
+
+        return $this;
+    }
+
+    /**
+     * @return string
+     */
+    public function getVersion()
+    {
+        return $this->version;
+    }
+
+    /**
+     * Loads the icon information from the composer.json.
+     *
+     * @return array|null
+     */
+    public function getIcon()
+    {
+        if (($icon = $this->composerJsonAttribute('extra.flarum-extension.icon'))) {
+            if ($file = Arr::get($icon, 'image')) {
+                $file = $this->path . '/' . $file;
+
+                if (file_exists($file)) {
+                    $mimetype = pathinfo($file, PATHINFO_EXTENSION) === 'svg'
+                        ? 'image/svg+xml'
+                        : finfo_file(finfo_open(FILEINFO_MIME_TYPE), $file);
+                    $data     = file_get_contents($file);
+
+                    $icon['backgroundImage'] = 'url(\'data:' . $mimetype . ';base64,' . base64_encode($data) . '\')';
```

Comment:
@franzliedke maybe push this to a seperate item? This pr is already monstruous and keeps all previous functionality _intact_.